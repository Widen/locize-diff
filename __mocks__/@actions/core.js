export const mocks = {
  projectId: jest.fn(),
  projectSlug: jest.fn(),
  apiKey: jest.fn(),
  leftVersion: jest.fn(),
  rightVersion: jest.fn(),
  includeDrafts: jest.fn(),
  ignoreDeletedKeys: jest.fn(),
}

const defaults = {
  projectId: 'projectId',
  apiKey: 'apiKey',
  leftVersion: 'latest',
  rightVersion: 'production',
  includeDrafts: 'false',
  ignoreDeletedKeys: 'true',
}

export function getInput(key) {
  return mocks[key]?.() ?? defaults[key]
}

export const setFailed = jest.fn()
