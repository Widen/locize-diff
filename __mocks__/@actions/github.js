export const listFilesMock = jest.fn()
export const createCommentMock = jest.fn()
export const updateCommentMock = jest.fn()
export const listCommentsMock = jest.fn()
export const graphqlMock = jest.fn()

const wrapResponse = () => async () => ({
  data: listCommentsMock() ?? [],
})

export const getOctokit = () => ({
  graphql: graphqlMock,
  issues: {
    createComment: createCommentMock,
    listComments: wrapResponse(listCommentsMock),
    updateComment: updateCommentMock,
  },
  pulls: {
    listFiles: wrapResponse(listFilesMock),
  },
})

export const pr = {
  eventName: 'pull_request',
  issue: {
    number: 123,
  },
}

export const prComment = (body) => ({
  eventName: 'issue_comment',
  actor: 'somebody',
  issue: {
    number: 123,
  },
  payload: {
    comment: {
      id: 492700400,
      node_id: 'MDEyOklzc3VlQ29tbWVudDQ5MjcwMDQwMA==',
      body,
    },
  },
})

export const contextMock = jest.fn()
export const context = new Proxy(
  {},
  { get: (_, property) => contextMock()[property] }
)
