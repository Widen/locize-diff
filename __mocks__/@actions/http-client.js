export const listResourcesMock = jest.fn()
export const fetchResourceMock = jest.fn()
export const putMock = jest.fn()

export class HttpClient {
  async get(url) {
    return {
      readBody: async () => {
        const mock = url.includes('/download/')
          ? listResourcesMock
          : url.includes('/pull/')
          ? fetchResourceMock
          : null

        return JSON.stringify(mock(url))
      },
    }
  }

  put = putMock
}
