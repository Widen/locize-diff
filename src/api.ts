import { getInput } from '@actions/core'
import { HttpClient } from '@actions/http-client'
import { BearerCredentialHandler } from '@actions/http-client/auth'
import { Dict, ResourceCollection } from './utils/types'

const authHandler = new BearerCredentialHandler(getInput('apiKey'))
const client = new HttpClient(undefined, [authHandler])

export function getJSON<T>(url: string): Promise<T> {
  return client
    .get(url)
    .then((res) => res.readBody())
    .then((res) => JSON.parse(res))
}

interface ListResourcesResponse {
  key: string
  lastModified: string
  size: number
  url: string
}

export async function listResources(
  projectId: string,
  version: string
): Promise<ListResourcesResponse[]> {
  return getJSON(`https://api.locize.app/download/${projectId}/${version}`)
}

export async function collectResources(
  projectId: string,
  version: string
): Promise<ResourceCollection[]> {
  const resources = await listResources(projectId, version)
  const promises = resources.map(async (resource) => {
    const [language, namespace] = resource.url.split('/').slice(-2)
    const url = resource.url.replace(projectId, `pull/${projectId}`)

    return {
      key: `${language}/${namespace}`,
      resources: await getJSON<Dict>(url),
    }
  })

  return Promise.all(promises)
}

export async function updateTranslations(
  key: string,
  updates: Record<string, string | null>
) {
  const projectId = getInput('projectId')
  const version = getInput('rightVersion')

  console.log(
    'Updating',
    `https://api.locize.app/update/${projectId}/${version}/${key}`,
    updates
  )

  await client.post(
    `https://api.locize.app/update/${projectId}/${version}/${key}`,
    JSON.stringify(updates)
  )
}
