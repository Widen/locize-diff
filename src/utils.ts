import { getJSON, listResources } from './api'
import { Dict, ResourceCollection } from './types'

export async function collectResources(
  projectId: string,
  version: string
): Promise<ResourceCollection[]> {
  const resources = await listResources(projectId, version)
  const promises = resources.map(async (resource) => {
    const [language, namespace] = resource.url.split('/').slice(-2)

    return {
      key: `${language}/${namespace}`,
      resources: await getJSON<Dict>(resource.url),
    }
  })

  return Promise.all(promises)
}
