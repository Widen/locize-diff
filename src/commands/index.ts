import { context } from '@actions/github'
import { createComment } from '../utils/comments'
import { runDiff } from './check'
import { runCopy } from './copy'

async function reportResult(message?: string) {
  if (message) {
    await createComment(`@${context.actor} ${message}`)
  }
}

export async function runCommand() {
  const comment = context.payload.comment?.body ?? ''
  const match = comment.match(/@locize-diff\s(check|copy)/)
  const command = match?.[1]

  switch (command) {
    case 'check':
      return reportResult(await runDiff())

    case 'copy':
      return reportResult(await runCopy())
  }
}
