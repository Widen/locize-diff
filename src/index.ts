import { getInput, setFailed } from '@actions/core'
import { context, getOctokit } from '@actions/github'
import { getComment, minimizeComment } from './comments'
import { diffResources } from './diff'
import { createMessage } from './message'
import { collectResources } from './utils'

async function main() {
  const projectId = getInput('projectId')
  const leftVersion = getInput('leftVersion')
  const rightVersion = getInput('rightVersion')
  const includeDrafts = getInput('includeDrafts') !== 'true'

  if (includeDrafts && context.payload.pull_request?.draft) {
    return
  }

  try {
    const octokit = getOctokit(getInput('token'))
    const left = await collectResources(projectId, leftVersion)
    const right = await collectResources(projectId, rightVersion)
    const diffs = diffResources(left, right)
    const comment = await getComment()

    if (diffs.length) {
      const req = {
        ...context.repo,
        body: createMessage(diffs),
        issue_number: context.issue.number,
      }

      if (comment) {
        if (comment.body !== req.body) {
          await octokit.issues.updateComment({ ...req, comment_id: comment.id })
        }
      } else {
        await octokit.issues.createComment(req)
      }
    }

    // If the comment exists and there are no longer any diffs, we minimize the
    // comment so it no longer shows in the GitHub UI.
    if (comment && !diffs.length) {
      await minimizeComment(comment.node_id)
    }
  } catch (err) {
    setFailed(err.message)
  }
}

main()
