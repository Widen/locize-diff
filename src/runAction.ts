import { setFailed } from '@actions/core'
import { context } from '@actions/github'
import { runCommand } from './commands'
import { runDiff } from './commands/diff'

export async function runAction() {
  try {
    if (context.eventName === 'issue_comment') {
      await runCommand()
    } else {
      await runDiff()
    }
  } catch (err) {
    setFailed(err.message)
  }
}
