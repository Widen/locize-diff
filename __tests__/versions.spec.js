jest.mock('@actions/core')
jest.mock('@actions/github')
jest.mock('@actions/http-client')

import { mocks } from '@actions/core'
import { createCommentMock } from '@actions/github'
import { listResourcesMock } from '@actions/http-client'
import { runAction } from '../src/runAction'
import { mockFetchResource } from './utils'

beforeEach(() => {
  jest.resetAllMocks()
})

it.only('should allow customizing the left and right versions', async () => {
  mocks.projectId.mockReturnValue('project-id')
  mocks.leftVersion.mockReturnValue('left')
  mocks.rightVersion.mockReturnValue('right')

  listResourcesMock.mockImplementation((url) => {
    const version = url.includes('/left') ? 'left' : 'right'
    const key = `project-id/${version}/es/common`

    return [{ url: `https://api.locize.app/${key}`, key }]
  })

  mockFetchResource(
    { 'es/common': { foo: 'bar' } },
    { 'es/common': { foo: 'baz' } }
  )

  await runAction()
  expect(createCommentMock).toHaveBeenCalledTimes(1)
  expect(createCommentMock.mock.calls[0][0].issue_number).toBe(123)
  expect(createCommentMock.mock.calls[0][0].body).toMatchSnapshot()
})
