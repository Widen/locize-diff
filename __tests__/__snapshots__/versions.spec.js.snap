// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should allow customizing the left and right versions 1`] = `
"Heads up! Looks like there are some differences between your two Locize versions.

### \`es/common\`

| Key | \`left\` | \`right\` |
| --- | --- | --- |
| \`foo\` | bar | baz |

_Generated by [Locize Diff](https://github.com/Widen/locize-diff)._"
`;
